server/sonar-server/src/main/java/org/sonar/server/test/index/TestIndexer.java

```diff
@@ -0,0 +1,123 @@
+import org.elasticsearch.index.query.QueryBuilders;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.es.BaseIndexer;
+import org.sonar.server.es.BulkIndexer;
+import org.sonar.server.es.EsClient;
+import org.sonar.server.source.index.FileSourcesUpdaterUtil;
+
+import javax.annotation.Nullable;
+import java.sql.Connection;
+import java.util.Iterator;
+
+import static org.sonar.server.source.index.SourceLineIndexDefinition.FIELD_PROJECT_UUID;
+import static org.sonar.server.test.index.TestIndexDefinition.*;
+
+/**
+ * Add to Elasticsearch index {@link TestIndexDefinition} the rows of
+ * db table FILE_SOURCES that are not indexed yet
+ */
+public class TestIndexer extends BaseIndexer {
+
+  private final DbClient dbClient;
+
+  public TestIndexer(DbClient dbClient, EsClient esClient) {
+    super(esClient, 0L, INDEX, TYPE, FIELD_UPDATED_AT);
+    this.dbClient = dbClient;
+  }
+
+  public void index(final String projectUuid) {
+    super.index(new IndexerTask() {
+      @Override
+      public long index(long lastUpdatedAt) {
+        return doIndex(lastUpdatedAt, projectUuid);
+      }
+    });
+  }
+
+  @Override
+  protected long doIndex(long lastUpdatedAt) {
+    return doIndex(lastUpdatedAt, null);
+  }
+
+  private long doIndex(long lastUpdatedAt, @Nullable String projectUuid) {
+    final BulkIndexer bulk = new BulkIndexer(esClient, INDEX);
+    bulk.setLarge(lastUpdatedAt == 0L);
+
+    DbSession dbSession = dbClient.openSession(false);
+    Connection dbConnection = dbSession.getConnection();
+    try {
+      TestResultSetIterator rowIt = TestResultSetIterator.create(dbClient, dbConnection, lastUpdatedAt, projectUuid);
+      long maxUpdatedAt = doIndex(bulk, rowIt);
+      rowIt.close();
+      return maxUpdatedAt;
+
+    } finally {
+      dbSession.close();
+    }
+  }
+
+  public long index(Iterator<FileSourcesUpdaterUtil.Row> dbRows) {
+    BulkIndexer bulk = new BulkIndexer(esClient, INDEX);
+    return doIndex(bulk, dbRows);
+  }
+
+  private long doIndex(BulkIndexer bulk, Iterator<FileSourcesUpdaterUtil.Row> dbRows) {
+    long maxUpdatedAt = 0L;
+    bulk.start();
+    while (dbRows.hasNext()) {
+      FileSourcesUpdaterUtil.Row row = dbRows.next();
+      for (UpdateRequest updateRequest : row.getUpdateRequests()) {
+        bulk.add(updateRequest);
+      }
+      maxUpdatedAt = Math.max(maxUpdatedAt, row.getUpdatedAt());
+    }
+    bulk.stop();
+    return maxUpdatedAt;
+  }
+
+  public void deleteByFile(String fileUuid) {
+    // TODO would be great to have the projectUuid for routing
+    SearchRequestBuilder searchRequest = esClient.prepareSearch(INDEX)
+      .setTypes(TYPE)
+      .setQuery(QueryBuilders.filteredQuery(
+        QueryBuilders.matchAllQuery(),
+        FilterBuilders.termFilter(FIELD_FILE_UUID, fileUuid).cache(false)));
+    BulkIndexer.delete(esClient, INDEX, searchRequest);
+  }
+
+  public void deleteByProject(String projectUuid) {
```

Comment:
This method is never used, but from what I see in SourceLineIndexer, shouldn't this method be called in ComponentCleanerService.deleteFromIndices() ?